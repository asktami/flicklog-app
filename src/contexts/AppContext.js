import React, { Component } from 'react';

const AppContext = React.createContext({
	movie: [],
	videos: [],
	reviews: [],
	movieList: [],
	watchList: [],
	reviewList: [],
	setPageName: () => {},
	setPageNumber: () => {},
	setTotalPages: () => {},
	setLoginUserId: () => {},
	setError: () => {},
	setQuery: () => {},
	clearError: () => {},
	setMovieList: () => {},
	setWatchList: () => {},
	setReviewList: () => {},
	setMovie: () => {},
	setVideos: () => {},
	setReviews: () => {},
	clearMovie: () => {},
	addReview: () => {},
	editReview: () => {},
	deleteReview: () => {},
	addWatchListItem: () => {},
	removeWatchListItem: () => {},
	showBackToTop: () => {},
	handleInputChange: () => {},
});
export default AppContext;

export class AppProvider extends Component {
	state = {
		loginUserId: '',
		error: null,
		query: '',
		movieList: [],
		watchList: [],
		reviewList: [],
		movie: [],
		videos: [],
		reviews: [],
		showBackToTop: false,
		pageName: '',
		pageNumber: 1,
		totalPages: 0,
	};

	setShowBackToTop = (show) => {
		this.setState({ showBackToTop: show });
	};

	setPageName = (pageName) => {
		this.setState({ pageName });
	};

	setPageNumber = (pageNumber) => {
		this.setState({ pageNumber });
	};

	setTotalPages = (totalPages) => {
		this.setState({ totalPages });
	};

	setLoginUserId = (id) => {
		this.setState({ loginUserId: id });
	};

	setError = (error) => {
		this.setState({ error });
	};

	clearError = () => {
		this.setState({ error: null });
	};

	setQuery = (q) => {
		this.setState({ query: q });
	};

	handleInputChange = async (e) => {
		await this.setQuery(e.target.value.trim());
	};

	setMovieList = (movieList) => {
		this.setState({ movieList });
	};

	setWatchList = (watchList) => {
		this.setState({ watchList });
	};

	setReviewList = (reviewList) => {
		this.setState({ reviewList });
	};

	setMovie = (movie) => {
		this.setState({ movie });
	};

	setVideos = (videos) => {
		this.setState({ videos });
	};

	setReviews = (reviews) => {
		this.setState({ reviews });
	};

	clearMovie = () => {
		this.setMovie([]);
		this.setReviews([]);
	};

	addReview = (review) => {
		this.setReviews([...this.state.reviews, review]);
	};

	editReview = (updatedReview) => {
		const newReviews = this.state.reviews.map((review) =>
			review.id !== updatedReview.id ? review : updatedReview
		);

		this.setReviews(newReviews);
	};

	deleteReview = (reviewId) => {
		const newReviews = this.state.reviews.filter(
			(review) => review.id !== reviewId
		);

		this.setReviews(newReviews);
	};

	addWatchListItem = ({ movie }) => {
		this.setWatchList([...this.state.watchList, movie]);
	};

	removeWatchListItem = (movieId) => {
		const newWatchList = this.state.watchList.filter(
			(watchlist) => watchlist.movie_id !== movieId
		);
		this.setWatchList(newWatchList);
	};

	render() {
		const value = {
			totalPages: this.state.totalPages,
			setTotalPages: this.setTotalPages,
			pageNumber: this.state.pageNumber,
			setPageNumber: this.setPageNumber,
			pageName: this.state.pageName,
			setPageName: this.setPageName,
			setLoginUserId: this.setLoginUserId,
			loginUserId: this.state.loginUserId,
			error: this.state.error,
			setError: this.setError,
			clearError: this.clearError,
			query: this.state.query,
			setQuery: this.setQuery,
			movieList: this.state.movieList,
			watchList: this.state.watchList,
			reviewList: this.state.reviewList,
			setMovieList: this.setMovieList,
			setWatchList: this.setWatchList,
			setReviewList: this.setReviewList,
			movie: this.state.movie,
			videos: this.state.videos,
			reviews: this.state.reviews,
			setMovie: this.setMovie,
			setVideos: this.setVideos,
			setReviews: this.setReviews,
			clearMovie: this.clearMovie,
			addReview: this.addReview,
			editReview: this.editReview,
			deleteReview: this.deleteReview,
			addWatchListItem: this.addWatchListItem,
			removeWatchListItem: this.removeWatchListItem,
			showBackToTop: this.showBackToTop,
			handleInputChange: this.handleInputChange,
		};
		return (
			<AppContext.Provider value={value}>
				{this.props.children}
			</AppContext.Provider>
		);
	}
}
