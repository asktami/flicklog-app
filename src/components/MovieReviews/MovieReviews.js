import React, { Component } from 'react';
import AppContext from '../../contexts/AppContext';
import MovieApiService from '../../services/movie-api-service';
import { Link } from 'react-router-dom';
import StarRating from '../StarRating/StarRating';

export default class MovieReviews extends Component {
	static contextType = AppContext;

	componentDidMount() {
		this.context.clearError();
	}

	handleClickDeleteReview = (review_id) => {
		this.context.clearError();

		MovieApiService.deleteReview(review_id)
			.then(() => this.context.deleteReview(review_id))
			.catch(this.context.setError);

		// test error
		// this.context.setError(
		// 	'Testing Error in MovieReviews - handleClickDeleteReview'
		// );
	};

	render() {
		const { error, loginUserId, reviews } = this.context;

		return (
			<>
				<ul className="review-list">
					<h3 className="movie-reviews-title">Reviews</h3>

					{error ? (
						<p className="error">
							There was an error, please wait a few minutes and try again.
							<br />
							{error.message
								? error.message
								: JSON.parse(JSON.stringify(error))}
						</p>
					) : null}

					{reviews.length === 0 && (
						<li className="review-item">
							<div className="review-text">No reviews.</div>
						</li>
					)}

					{(reviews || []).map((review) => (
						<li key={review.id} className="review-item">
							<div className="review-text">
								{review.review}
								<br />

								<div className="flex-row review-footer ">
									<div>
										<StarRating rating={review.rating} />
										<br />
										<span className="review-user">{review.fullname}</span>
									</div>
									{review.user_id === loginUserId ? (
										<div className="btn-container">
											<Link to={`/reviews/${review.id}`}>
												<button className="btn btn-as-link" type="submit">
													Edit
												</button>
											</Link>
											&nbsp;&nbsp;&nbsp;
											<button
												className="btn btn-as-link"
												onClick={() => this.handleClickDeleteReview(review.id)}
											>
												Delete
											</button>
										</div>
									) : null}
								</div>
							</div>
						</li>
					))}
				</ul>
			</>
		);
	}
}
