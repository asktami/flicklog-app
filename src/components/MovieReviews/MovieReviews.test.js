import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';


import AppContext from '../../contexts/AppContext';
import MovieReviews from './MovieReviews';

describe(`MovieReviews component`, () => {
	const props = {
		match: {
			params: {
				movie_id: 348
			}
		}
	};
	

	it('renders the MovieReviews given context', () => {
	
	const wrapper = shallow(
    <AppContext.Provider
    value={{
      dispatch: jest.fn(),
      reviews:  [
			{
				id: 1,
				review: 'Alien test review!',
				rating: 4,
				date_created: '2019-10-31 23:14:12.649275',
				movie_id: '348',
				user_id: 1
			},
			{
				id: 2,
				review: 'Star Wars test review!',
				rating: 4,
				date_created: '2019-10-31 23:14:12.649275',
				movie_id: '11',
				user_id: 1
			}
		]
    }}>
      <MovieReviews />
    </AppContext.Provider>
    )
    
    
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
