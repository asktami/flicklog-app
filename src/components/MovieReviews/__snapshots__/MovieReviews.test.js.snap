// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieReviews component renders the MovieReviews given context 1`] = `<MovieReviews />`;
